pageserver/src/tenant/storage_layer/layer_contents/virtual_value.rs

```diff
@@ -0,0 +1,207 @@
+                                image: img,
+                                lsns: vec![new_lsn],
+                                records: vec![new_rec],
+                            },
+                            VirtualValue::NaturalWalRecord(vv_start) => {
+                                if vv_start.will_init() {
+                                    VirtualValue::ClosedRecLineage {
+                                        image_rec: vv_start,
+                                        lsns: vec![new_lsn],
+                                        records: vec![new_rec],
+                                    }
+                                } else {
+                                    VirtualValue::OpenLineage {
+                                        lsns: vec![new_lsn],
+                                        records: vec![vv_start, new_rec],
+                                    }
+                                }
+                            }
+                            VirtualValue::ClosedLineage {
+                                image,
+                                mut lsns,
+                                mut records,
+                            } => {
+                                lsns.push(new_lsn);
+                                records.push(new_rec);
+
+                                VirtualValue::ClosedLineage {
+                                    image,
+                                    lsns,
+                                    records,
+                                }
+                            }
+                            VirtualValue::ClosedRecLineage {
+                                image_rec,
+                                mut lsns,
+                                mut records,
+                            } => {
+                                lsns.push(new_lsn);
+                                records.push(new_rec);
+
+                                VirtualValue::ClosedRecLineage {
+                                    image_rec,
+                                    lsns,
+                                    records,
+                                }
+                            }
+                            VirtualValue::OpenLineage {
+                                mut lsns,
+                                mut records,
+                            } => {
+                                lsns.push(new_lsn);
+                                records.push(new_rec);
+
+                                VirtualValue::OpenLineage { lsns, records }
+                            }
+                        };
+
+                        self.state = Some((start_lsn, new_vv));
+
+                        None
+                    }
+                }
+            }
+        }
+    }
+
+    pub fn finish(mut self) -> Option<(Lsn, VirtualValue)> {
+        self.state.take()
+    }
+}
+
+impl Drop for VirtualValueBuilder {
+    fn drop(&mut self) {
+        assert!(self.state.is_none());
```

Comment:
Builder values usually don't do this to enable cancellation of whatever they are building.

This assumption would be important to note on the VirtualValueBuilder and/or `finish` method.

Reply:
First builder that comes to mind is from tokio, which does nothing unless `build` is called: https://docs.rs/tokio/latest/tokio/runtime/struct.Builder.html or std's https://doc.rust-lang.org/stable/std/fs/struct.OpenOptions.html.

I guess the separation could come if the builder would retain an `&mut` reference to something which could be left in an inconsistent state, which is not the case with this builder.

Perhaps this is as close as it gets: https://rust-unofficial.github.io/patterns/patterns/creational/builder.html

More official mentions in the rust api guidelines, which were fleshed out to get some crates to 1.0.0: https://rust-lang.github.io/api-guidelines/type-safety.html?highlight=builder#builders-enable-construction-of-complex-values-c-builder

But perhaps here is no strict "don't do this" but just that it is surprising to find such.